### ISSUE TITLE
Implement constants of actions as different data type (e.g. enum class instead of std::string) 

### ISSUE THREAD

**Issue by ajaust** [precice/precice#715]
I propose the change of the API for clearer and cleaner to move from `std::string` as identifier for actions to `enum class` or something similar in [`src/cplscheme/Constants.cpp`](https://github.com/precice/precice/blob/bc27c26d195185e61663f580d469399c6587010f/src/cplscheme/Constants.cpp). 

Major improvements:
- It limits the number of available actions in a "natural" way. An `enum class` can only take certain, pre-defined values while the `std::string` could take arbitrary values of arbitrary length.
- It separates description of what it is (an action) and what it does (writes initial data, e.g.) in a clean way. The `enum class` is called `Actions` and its members/values describe the actual action performed/checked for.
- Minimizes risk of forgetting to account for an action. One can use the `switch` statement and the compiler has the chance to warn you about handing other actions. When using the definition of `Actions` mentioned below the following code would produce a compiler warning:
```c++
bool isActionRequired( const constants::Actions action) const {
  switch (action)  {
    case Actions::writeInitialData:  {  //do something  }
    //Oh no! We forgot to handle other actions.
  }
}
```
Minor improvements:
- Comparison should be cheaper than for strings (probably not a performance concern at the moment)
- For me personally it feels more correct in C++ to use `enum class`.
Drawback:
It will break compatibility with the API and thus is probably something for preCICE 3.0. However, one could introduce it in a future release to enable both, the `std::string`-based and the `enum class`-based version and phase out the old approach slowly.

The solution I would propose (I am playing with it myself at the moment) would look like that:
```c++
enum class Actions {
  writeInitialData,
  writeIterationCheckpoint,
  readIterationCheckpoint
};
```

**Comment by uekerman** [comment]
Thanks, @ajaust! I agree in general, but one pitfall will be the bindings, which use `SolverInterface` and do only know about C features, right? Or is there a classical solution? Of course, this could be a feature of a "new" additional C++ interface, see #717. Then, it would neither break compatibility.

**Comment by ajaust** [comment]
I guess one solution would be to use enumerations `enum` which are a C feature instead of strong typed enumerations `enum class`. Another solution could be to use the strong typed enumerations in the (potentially new) C++ interface and internally still use the C interface. Honestly, I am not sure what would be better and easier to maintain.

**Comment by fsimonis** [comment]
@ajaust I totally agree with your standpoint.
Currently we require our C and Fortran users to juggle with `const char*`, which is a bit crazy.

Extending your suggestion with an underlying type we can use these values in the C/Fortran bindings.
```cpp
enum class Actions : int {
  writeInitialData,
  writeIterationCheckpoint,
  readIterationCheckpoint
};
```

This allows us to declare functions in c and fortran that can forward the underlying enum values:

C++ | C & Fortran (with `f`) function
--- | ---
`Action::writeInitialData` | `int precicec_action_writeInitialData()`
`Action::writeIterationCheckpoint` | `int precicec_action_writeIterationCheckpoint()`
`Action::readIterationCheckpoint` | `int precicec_action_readIterationCheckpoint()`

We then require `Action`/`int` overloads for both action methods:
```cpp
void SolverInterface::markActionFulfilled(precice::Action action);
void precicec_markActionFulfilled(int action);
void precicec_markActionFulfilled(int action);
```
We could also add a `typedef precice_action_t int;` for C and Fortran to extend the above functions.

The string version in the C++ SolverInterface can the be implemented in terms of the new interface and marked as deprecated.

---

An alternative approach would be to use explicit calls as proposed in issue https://github.com/precice/precice/issues/719.

**Comment by fsimonis** [comment]
This will be solved by #719
